File: core.buildtime/src/main/java/com/cloudibpm/core/data/HandwritingConstant.java
/**
 * 
 */
package com.cloudibpm.core.data;

import com.cloudibpm.core.buildtime.wfprocess.WfProcess;
import com.cloudibpm.core.util.encode.Base64Util;

/**
 * @author Dahai Cao created on 2017-11-27
 *
 */
public class HandwritingConstant extends Constant {

	/**
	 * serialVersionUID
	 */
	private static final long serialVersionUID = -7974915660546098677L;

	/**
	 * 
	 */
	public HandwritingConstant() {
		this.setName("空写字板对象"); // signature name
		this.setValue("");
		this.setDatatype(DataType.HANDWRITING);
	}

	@Override
	public Object clone(WfProcess owner) {
		HandwritingConstant d = new HandwritingConstant();
		d.setValue(this.getValue());
		return d;
	}

	@Override
	public boolean equals(Object obj) {
		if (obj instanceof HandwritingConstant) {
			if ((this.getValue() == null && ((HandwritingConstant) obj).getValue() == null)
					|| (this.getValue() != null && this.getValue().equals(((HandwritingConstant) obj).getValue()))) {
				return true;
			}
		} else if (obj instanceof String) {// it is used to recognize Expression
			if (this.getValue() != null && this.getValue().equals(obj)) {
				return true;
			}
		}
		return false;
	}

	@Override
	public String toExpressionString() {
		return "C@" + this.getDatatype() + "@" + this.getName() + "@" + Base64Util.encode(this.getValue());
	}

	@Override
	public void parseString(String str) {
		String[] ary = str.split("@");
		this.setDatatype(ary[1]);
		this.setName(ary[2]);
		if (ary.length > 2)
			this.setValue(Base64Util.decode(ary[3]));
	};

	@Override
	public String toString() {
		return this.getName();
	}

}

File: pm/src/main/webapp/js/data/handwritingconstant.js
/**
 * 
 */

function HandwritingConstant() {
	this.name = "空写字板对象"; // signature name
	this.value = "";
	this.datatype = "Handwriting";
};

HandwritingConstant.prototype = new Constant();

HandwritingConstant.prototype.clone = function(owner) {
	var d = new HandwritingConstant();
	d.name = this.name;
	d.value = this.value;
	return d;
};

HandwritingConstant.prototype.toExpressionString = function() {
	return "C@" + this.datatype + "@" + this.name + "@"
			+ Base64.encode(this.value);
};

HandwritingConstant.prototype.parseString = function(str) {
	var ary = str.split("@");
	this.name = ary[1];
	this.datatype = ary[2];
	this.value = Base64.decode(ary[3]);
};

HandwritingConstant.prototype.parseObject = function(o) {
	this.name = o.name;
	this.datatype = o.datatype;
	this.value = o.value;
};

// 5 Canvas 如何自适应屏幕大小
// http://lvxineye.iteye.com/blog/1671080
HandwritingConstant.prototype.toString = function() {
	return this.name;
};